synapse/storage/databases/main/event_push_actions.py

```diff
@@ -444,6 +445,20 @@ def _get_thread(thread_id: str) -> NotifCounts:
                 return main_counts
             return thread_counts.setdefault(thread_id, NotifCounts())
 
+        receipt_types_clause, receipts_args = make_in_list_sql_clause(
+            self.database_engine,
+            "receipt_type",
+            (ReceiptTypes.READ, ReceiptTypes.READ_PRIVATE),
+        )
+
+        # PostgreSQL and SQLite differ in comparing scalar numerics.
+        if isinstance(self.database_engine, PostgresEngine):
+            # GREATEST ignores NULLs.
+            receipt_stream_clause = "GREATEST(receipt_stream_ordering, ?)"
+        else:
+            # MAX returns NULL if any are NULL, so COALESCE to 0 first.
+            receipt_stream_clause = "MAX(COALESCE(receipt_stream_ordering, 0), ?)"
```

Comment:
We should add a comment to explain what some of these `?` are. It's hard to follow given there is a big disconnect between where this clause is used.

It looks like this is using the `receipt_stream_ordering` arg as well so `GREATEST(receipt_stream_ordering, /* receipt_stream_ordering */ ?)` but that is confusing since the arg and query alias name are the same.

We should probably differentiate them somehow. Not sure exactly the difference to give a good suggestion.

Reply:
I expanded the comment a bit in 162bd8dcdf23c7a2b409396acba330fb3b550698 and renamed some variables to clarify this.

Reply:
👍 Looks like it also moved to `unthreaded_receipt_stream_ordering ` and `threaded_receipt_stream_ordering` which makes it a lot better too